### ISSUE TITLE
High I/O consumption on wazuh-clusterd when multiple agents re-register

### ISSUE THREAD

**Issue by francobep** [wazuh/wazuh#6035]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| v3.12.3 | wazuh-clusterd | Manager | docker images | EKS/amazon-linux |

When many (+100 per hour) agents are re-registered, we see an increment of the I/O access (write) from the wazuh-clusterd process.

![image](https://user-images.githubusercontent.com/632969/93341061-0bd09c80-f804-11ea-9929-a3357fb30b9e.png)

Trace the volume virtual device /dev/nvme4n1 as root user on the instance:

```
[root@instance ~]# timeout 300 blktrace -d /dev/nvme4n1 -o - | blkparse -i - > nvme4n1.trace

[root@instance ~]# wc -l nvme4n1.trace
153198 nvme4n1.trace

```
We can see two processes with i/o (write)

- remoted
- python ( wazuh-clusterd )

```
[root@instance ~]# 
[root@instance ~]# grep -v jbd2 nvme4n1.trace| grep python | wc -l
256050
[root@instance ~]# grep -v jbd2 nvme4n1.trace| grep remoted | wc -l
870
[root@instance ~]# wc -l nvme4n1.trace
334916 nvme4n1.trace
```

Tracing the `wazuh-clusterd` process with `strace`:

```
[root@instance ~]# timeout 300 strace -f -y -e trace=read,write -o 1586.trace -p 1586

[root@instance ~]# grep write 1586.trace | wc -l
241074

[root@instance ~]# grep write 1586.trace | grep zip | wc -l
189942

```

`clusterd` is writing many times per second a lot of files like:

```
17</var/ossec/queue/cluster/wazuh-manager-worker-1/wazuh-manager-worker-1-1599715699.8472981-7827462739174045.zipdir/cluster_control.json

```

In `cluster.log` on the managers, we see this error repeated many times:

```
2020/09/10 14:00:43 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] [Cluster] Error 3001 - Error creating zip file: [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/117137'
2020/09/10 14:00:43 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-2] [Main] [Cluster] Error 3001 - Error creating zip file: [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/117137'
2020/09/10 14:00:52 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] [Cluster] Error 3001 - Error creating zip file: [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/117148'
2020/09/10 14:00:52 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-2] [Main] [Cluster] Error 3001 - Error creating zip file: [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/117148'
2020/09/10 14:00:57 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-1] [Main] [Cluster] Error 3001 - Error creating zip file: [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/117151'
2020/09/10 14:01:01 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] [Cluster] Error 3001 - Error creating zip file: [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/117151'

```

**Comment by Selutario** [comment]
## Issue update

Hey team,

I have just started researching the problem reported here. There is nothing remarkable yet that I can point out, but it would be nice to point out a few things. Some of the logs that can be found in workers are similar to this:
```
2020/09/10 00:00:09 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] Error processing missing file '/queue/agent-groups/115778': [Errno 2] No such file or directory: '/var/ossec/queue/cluster/wazuh-manager-worker-0/wazuh-manager-worker-0-1599696009.1553571-3655500237389955.zipdir/queue/agent-groups/115778'
2020/09/10 00:00:09 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] Error processing missing file '/queue/agent-groups/112290': [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/112290'
2020/09/10 00:00:09 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] Error processing missing file '/queue/agent-groups/111293': [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/111293'
```

That exception is being raised in this place:
https://github.com/wazuh/wazuh/blob/d7d0a1a58e1729b4bb4ecfbfc9dbc66334b2dd3c/framework/wazuh/cluster/worker.py#L421-L452

The problem is occurring when trying to create the files that exist in the master and are missing in the worker. According to the content of the error, it could be thought that an attempt is being made to overwrite a file that does not exist, which causes an error when it is not found. Also the path of the first file seems weird compared to the other ones shown in the log. More research is still needed.

Regards,
Selu.

**Comment by Selutario** [comment]
## Issue update

Hi again team,

The error, as it seems obvious, is because the file which is trying to be moved with `safe_move()` does not exist. However, I have not yet been able to determine the origin of this solution.

### Assumption 1
The zip contains the `cluster_control.json` file, but not the file with the actual content of agent-groups, for example, `/queue/agent-groups/115776`.

This would cause the theoretical path of the file to be formed (using the information of `cluster_control.json`), but when trying to move it, it is not possible to do so since it is not included in the zip.

https://github.com/wazuh/wazuh/blob/d7d0a1a58e1729b4bb4ecfbfc9dbc66334b2dd3c/framework/wazuh/cluster/worker.py#L428

We should try to determine, in that case, why the zip does not include the corresponding file to be copied to the worker.

### Assumption 2

The file is included, but something deletes it before it is moved. In theory, this file is deleted by the integrity task itself once the update process is finished, as seen in line 300:

https://github.com/wazuh/wazuh/blob/d7d0a1a58e1729b4bb4ecfbfc9dbc66334b2dd3c/framework/wazuh/cluster/worker.py#L279-L300

This option can be discarded for now.

Best regards,
Selu.

**Comment by Selutario** [comment]
# Conclusion

Hi team!

After analyzing the complete behaviour of the integrity synchronization, we can **confirm that the problem lies in the simultaneous management of files by `ossec-remoted` and `wazuh-clusterd` services as a consequence of the constant re-registration of agents**.

## Erros in workers and master
I detail the workflow below to make it easier to understand the problem:

### 1 Calculate integrity
Both the master node and the workers periodically calculate (usually every 10 seconds) the integrity of their files. To do this, they go through the files in each directory specified within [cluster.json](https://github.com/wazuh/wazuh/blob/3.12/framework/wazuh/cluster/cluster.json) and obtain the modification date, the md5 hash and other data for each one. 

A dictionary is created that contains a key for each path of the found files, and whose value is the information of the file specified in the key.

### 2 Compare integrity
The workers send that dictionary with the integrity information to the master, who compares it with its own as shown in line 468.
https://github.com/wazuh/wazuh/blob/d7d0a1a58e1729b4bb4ecfbfc9dbc66334b2dd3c/framework/wazuh/cluster/master.py#L468

When doing the comparison, it finds that there are missing files in the worker, and then it generates a dictionary like the one below with all the file paths (keys) and info of each file (value):
```
{'missing': missing_files, 'extra': extra_files, 'shared': shared_files, 'extra_valid': extra_valid_files}
```

### 3. Compress files to send
The master tries to compress the missing and shared files (the ones it has to send to the worker). 
https://github.com/wazuh/wazuh/blob/d7d0a1a58e1729b4bb4ecfbfc9dbc66334b2dd3c/framework/wazuh/cluster/master.py#L477-L483
The [compress_files()](https://github.com/wazuh/wazuh/blob/d7d0a1a58e1729b4bb4ecfbfc9dbc66334b2dd3c/framework/wazuh/cluster/cluster.py#L206-L226) function receives a list of paths with all the files to be sent, and also the dictionary `worker_files_ko` with the integrity (each key (path) and value (info)) that is compressed under the filename of `cluster_control.json`.

### 4. Origin of the error
During the comparing and compression process, ossec-remoted may have deleted some of the files that were on the list of files to compress. It raises an exception that generates an error log like this, but does not stop the process:
```
2020/09/10 00:00:09 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] [Cluster] Error 3001 - Error creating zip file: [Errno 2] No such file or directory: '/var/ossec/queue/agent-groups/115778'
```
Finally, it also generates the `cluster_control.json` file with all the integrity information it had after the comparison. That is, the `cluster_control.json` which is sent to the worker has file paths that could not be compressed because they no longer existed.

### 5. Worker
The worker reads each of the paths that appear in that `cluster_control.json`, and tries to overwrite its file with the one that is supposed to be in the zip, but the path of the file it is looking for does not exist in the zip because it was not compressed. That causes this other error on the worker:
```
2020/09/10 00:00:09 wazuh-clusterd: ERROR: [Worker wazuh-manager-worker-0] [Main] Error processing missing file '/queue/agent-groups/115778': [Errno 2] No such file or directory: '/var/ossec/queue/cluster/wazuh-manager-worker-0/wazuh-manager-worker-0-1599696009.1553571-3655500237389955.zipdir/queue/agent-groups/115778'
```

## High use of resources
When re-registering an agent, the files associated with it are deleted and re-created. This causes a new modification date and other changes to these files, and therefore the integrity file synchronization process has to be performed again.

This process consumes both I/O and network resources. When done over multiple agents and repeatedly, it causes a great increase in all those resources. 

https://github.com/wazuh/wazuh/blob/d7d0a1a58e1729b4bb4ecfbfc9dbc66334b2dd3c/framework/wazuh/cluster/master.py#L474-L483

Without changes in integrity, so many resources are not consumed since it is not necessary to compress and send anything as seen in line 476.

## Fix
The solution is not clear. It would be necessary to prevent ossec-remoted from deleting files that are being used by wazuh-clusterd. This would avoid the errors shown in the log, but not the high resource usage that is inherent in the integrity synchronization process.

Best regards,
Selu.

**Comment by davidjiglesias** [comment]
# Status update

Hi team!

After analyzing the cluster environment we have reached the following conclusions on why the disk I/O consumption is so high, and other things to consider.


1.  The `master-node` had several thousands of agent-group files in `/var/ossec/queue/agent-groups` belonging to agents which are not currently registered. That error is related to https://github.com/wazuh/wazuh/issues/5438 and should be fixed for `Wazuh 4.0.0`.
2. The `master-node` tries to synchronize all those files (around 16000 when the investigation was made) to the `worker-nodes`. Creating huge zip files to synchronize them:
![worker2-zip](https://user-images.githubusercontent.com/48716145/97549283-80d0ed80-19d0-11eb-81e9-2bc1ad268699.png)
3. The `worker-nodes` receive these files from the `master-node` as `missing` files which they need to add to their `/var/ossec/queue/agent-groups` folder, however, in the worker nodes, when a file is created inside `/var/ossec/queue/agent-groups` folder, `wazuh-modulesd` immediately deletes the file if it doesn't belong to an agent that is registered in `client.keys`. For that reason, the files are not really synchronized and they will be send again by the master in the next integrity iteration:
```
2020/09/10 00:00:05 wazuh-clusterd: INFO: [Worker wazuh-manager-worker-0] [Main] Analyzing received files: Missing: 1104. Shared: 0. Extra: 0. ExtraValid: 0
2020/09/10 00:00:09 wazuh-clusterd: INFO: [Worker wazuh-manager-worker-0] [Main] Analyzing received files: Missing: 1102. Shared: 0. Extra: 0. ExtraValid: 0
2020/09/10 00:00:18 wazuh-clusterd: INFO: [Worker wazuh-manager-worker-0] [Main] Analyzing received files: Missing: 1100. Shared: 0. Extra: 0. ExtraValid: 0
2020/09/10 00:00:27 wazuh-clusterd: INFO: [Worker wazuh-manager-worker-0] [Main] Analyzing received files: Missing: 1099. Shared: 0. Extra: 0. ExtraValid: 0
```
4. The file permissions for `/var/ossec/queue/agent-groups` inside the `master-node` are weird, with the mayority of them being owned by `root`, and only those with `default` as group having `ossec` as owner:
![permissions agent-groups](https://user-images.githubusercontent.com/48716145/97550124-8b3fb700-19d1-11eb-8634-0ebeffaf304c.png)
The agent-groups files should only be handled by `ossec-remoted` (which creates the original file when the agent connect and adds default group to it), having owner `ossecr` and `wazuh-clusterd`, which synchronizes the files between nodes, having `ossec` as owner. Who is creating/modifying these files with `root` owner?
